### PR TITLE
Documentation of flow framework internals

### PR DESCRIPTION
**Pull request by exFalso**
This is by no means complete. Any topics that need expanding?

### CODE REVIEW

File: docs/source/contributing-flow-internals.rst [hunk 1]
```diff
@@ -0,0 +1,581 @@
...
+     s3()
+
+     return 1;
+   }
+
+   fun n1() { .. }
+   fun n2() { .. }
+   fun n3() { .. }
+   fun n4() { .. }
+   @Suspendable fun s2(b) { .. }
```
Comment (florian-f) [inline docs/source/contributing-flow-internals.rst]: What about s1?

File: docs/source/contributing-flow-internals.rst [hunk 2]
```diff
@@ -0,0 +1,581 @@
...
+data related to the management of the ``Fiber``, e.g. its state enum or identifier.
+
+The main try-catch that handles the yielding may be found `here <https://github.com/puniverse/quasar/blob/db0ac29f55bc0515023d67ab86a2178c5e6eeb94/quasar-core/src/main/java/co/paralleluniverse/fibers/Fiber.java#L790>`_.
+
+Checkpoints
+-----------
+
+The main idea behind checkpoints is to utilize the ``Fiber`` data structure and treat it as a serializable object capturing the state of a
+running computation. Whenever a Corda-suspendable API is hit, we capture the execution stack and corresponding entry list, and serialize
+it using Kryo. We thus get a handle to an arbitrary suspended computation.
```
Comment (florian-f) [inline docs/source/contributing-flow-internals.rst]: A link to Kryo would be nice here.

**Comment by gendal** [comment]
this is great!!  given that @florian-f will be on secondment to devrel for a period soon let's also make sure one or more of the other reviewers takes a good look at this before @exFalso departs so that we have somebody else who can ask lots of questions and get it into their heads too :)

File: docs/source/contributing-flow-internals.rst [hunk 3]
```diff
@@ -0,0 +1,588 @@
...
+^^^^^^^
+
+Handling of ``Event.Suspend`` is quite straightforward and is done `here <https://github.com/corda/corda/blob/26855967989557e4c078bb08dd528231d30fad8b/node/src/main/kotlin/net/corda/node/services/statemachine/transitions/TopLevelTransition.kt#L143>`_.
+We take the serialized ``Fiber`` and the IO request and create a new checkpoint, then depending on whether we should persist or not we
+either simply commit the database transaction and schedule a ``DoRemainingWork`` (to be explained later), or we persist the checkpoint, run
+the ``DeduplicationHandler`` inside-tx hooks, commit, then run the after-tx hooks, and schedule a ``DoRemainingWork``.
+
+Every checkpoint persistence implies the above steps, in this specific order.
+
+DoRemainingWork
```
Comment (rick-r3) [inline docs/source/contributing-flow-internals.rst]: There's a bunch of typos / formatting / word transpositions in this section.

File: docs/source/contributing-flow-internals.rst [hunk 4]
```diff
@@ -0,0 +1,588 @@
...
+                    currentState = currentState.copy(
+                            checkpoint = currentState.checkpoint.copy(
+                                    errorState = errorState.copy(propagating = true)
+                            )
+                    )
+                    actions.add(Action.ScheduleEvent(Event.DoRemainingWork))
+
+It marks the error state as ``propagating = true`` and schedules a ``DoRemainingWork``. The processing of that event in turn will detect
+that we are errored and propagating, and there are some errors that haven't been propagated yet. It then propagates those errors and updates
+the "propagated index" to indicate all errors have been dealt with. Subsequent ``DoRemainingWork``s will thus do nothing. However, in case
```
Comment (rick-r3) [inline docs/source/contributing-flow-internals.rst]: I think the ``s breaks the formatting

File: docs/source/contributing-flow-internals.rst [hunk 5]
```diff
@@ -0,0 +1,588 @@
...
+some other error condition or external event adds another error to the flow, we would automatically propagate that too, we don't need to
+write a special case for it.
+
+Most of the state machine logic is therefore about the handling ``DoRemainingWork``. Another example is resumptions due to an IO request
```
Comment (rick-r3) [inline docs/source/contributing-flow-internals.rst]: the handling of

**Comment by rick-r3** [comment]
Looks pretty good to me.

**Comment by rick-r3** [comment]
This has some enterprise mini-secrets in it.  Maybe we should move to enterprise?

**Comment by exFalso** [comment]
@rick-r3 the enterprise doctree seems to be very different, is that actively maintained? I can also just remove the dedupe part

File: docs/source/contributing-flow-internals.rst (hunk 3 above)
Comment (exFalso) [reply docs/source/contributing-flow-internals.rst]: Yes, the generated docs are outdated

**Pull request merged**